drivers/mdio/mdio_xlnx_gem.c

```diff
@@ -0,0 +1,302 @@
+		return -ETIMEDOUT;
+	}
+
+	/* Assemble & write the read command to the gem.phy_maint register */
+
+	/* Set the bits constant for any operation */
+	reg_val = ETH_XLNX_GEM_PHY_MAINT_CONST_BITS;
+	/* Indicate a read operation */
+	reg_val |= ETH_XLNX_GEM_PHY_MAINT_WRITE_OP_BIT;
+	/* PHY address */
+	reg_val |= (((uint32_t)prtad & ETH_XLNX_GEM_PHY_MAINT_PHY_ADDRESS_MASK)
+		    << ETH_XLNX_GEM_PHY_MAINT_PHY_ADDRESS_SHIFT);
+	/* Register address */
+	reg_val |= (((uint32_t)regad & ETH_XLNX_GEM_PHY_MAINT_REGISTER_ID_MASK)
+		    << ETH_XLNX_GEM_PHY_MAINT_REGISTER_ID_SHIFT);
+	/* 16 bits of data for the destination register */
+	reg_val |= ((uint32_t)data & ETH_XLNX_GEM_PHY_MAINT_DATA_MASK);
+
+	sys_write32(reg_val, dev_conf->gem_base_addr + ETH_XLNX_GEM_PHY_MAINTENANCE_OFFSET);
+
+	/*
+	 * Wait until gem.net_status[phy_mgmt_idle] == 1 -> current command
+	 * completed.
+	 */
+	poll_cnt = 0;
+	do {
+		if (poll_cnt++ > 0) {
+			k_busy_wait(CONFIG_MDIO_XLNX_GEM_POLL_DELAY);
+		}
+		reg_val = sys_read32(dev_conf->gem_base_addr + ETH_XLNX_GEM_NWSR_OFFSET);
+	} while ((reg_val & ETH_XLNX_GEM_NWSR_MDIO_IDLE_BIT) == 0 &&
+		 poll_cnt < CONFIG_MDIO_XLNX_GEM_MAX_POLL_RETRIES);
+	if (poll_cnt == CONFIG_MDIO_XLNX_GEM_MAX_POLL_RETRIES) {
+		LOG_ERR("%s: write to PHY address %hhu, register address %hhu timed out", dev->name,
+			prtad, regad);
+		return -ETIMEDOUT;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief GEM MDIO interface initialization function
+ * Configures the MDC clock divider in the associated GEM instance's
+ * net_config (NWCFG) register and sets the MDIO enable bit in the
+ * net_control (NWCTRL) register.
+ *
+ * @param dev Pointer to the GEM MDIO device
+ * @return    always returns 0
+ */
+static int xlnx_gem_mdio_initialize(const struct device *dev)
+{
+	const struct xlnx_gem_mdio_config *const dev_conf = dev->config;
+
+	uint32_t reg_val;
+	uint32_t mdc_divider = (uint32_t)MDC_DIVIDER_224;
+
+	/* Set the MDC divider in gem.net_config */
+	reg_val = sys_read32(dev_conf->gem_base_addr + ETH_XLNX_GEM_NWCFG_OFFSET);
+	reg_val &= ~(ETH_XLNX_GEM_NWCFG_MDC_MASK << ETH_XLNX_GEM_NWCFG_MDC_SHIFT);
+	reg_val |= ((mdc_divider & ETH_XLNX_GEM_NWCFG_MDC_MASK) << ETH_XLNX_GEM_NWCFG_MDC_SHIFT);
+	sys_write32(reg_val, dev_conf->gem_base_addr + ETH_XLNX_GEM_NWCFG_OFFSET);
+
+	/* Enable the MDIO interface */
+	reg_val = sys_read32(dev_conf->gem_base_addr + ETH_XLNX_GEM_NWCTRL_OFFSET);
+	reg_val |= ETH_XLNX_GEM_NWCTRL_MDEN_BIT;
+	sys_write32(reg_val, dev_conf->gem_base_addr + ETH_XLNX_GEM_NWCTRL_OFFSET);
+
+	LOG_DBG("%s: initialized", dev->name);
+	return 0;
+}
+
+static DEVICE_API(mdio, xlnx_gem_mdio_api) = {
+	.read = xlnx_gem_mdio_read,
+	.write = xlnx_gem_mdio_write,
+};
+
+#define XLNX_GEM_MDIO_DEV_CONFIG(port)                                                             \
+	static const struct xlnx_gem_mdio_config xlnx_gem##port##_mdio_cfg = {                     \
+		.gem_base_addr = DT_REG_ADDR_BY_IDX(DT_INST(port, xlnx_gem), 0),                   \
```

Comment:
if it needs the base address of the ethernet mac, it's a better way to have a common parent and the ethernet mac and mdio as a child of it. like here: 
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/ethernet/st%2Cstm32-ethernet-controller.yaml and 
https://github.com/zephyrproject-rtos/zephyr/blob/179045e5a570e5a6e978651ecd3c4e3f7a3a4a4e/dts/arm/st/f7/stm32f765.dtsi#L71-L94